packages/flutter_tools/test/base/context_test.dart

```diff
@@ -2,100 +2,260 @@
+          },
+        );
+        expect(value, isNull);
       });
 
-      expect(mockLogger.errorText, '');
-      expect(mockLogger.statusText, '');
-      expect(mockLogger.traceText, 'foo bar\n');
+      test('throws if generator has dependency cycle', () async {
+        final Future<String> value = context.run<Future<String>>(
+          body: () async {
+            return context[String];
+          },
+          fallbacks: <Type, Generator>{
+            int: () => int.parse(context[String]),
+            String: () => '${context[double]}',
+            double: () => context[int] * 1.0,
+          },
+        );
+        try {
+          await value;
+          fail('ContextDependencyCycleException expected but not thrown.');
+        } on ContextDependencyCycleException catch (e) {
+          expect(e.cycle, <Type>[String, double, int]);
+          expect(e.toString(), 'Dependency cycle detected: String -> double -> int');
+        }
+      });
     });
 
-    test('awaitNestedZones', () async {
-      final AppContext outerContext = new AppContext();
-      await outerContext.runInZone(() async {
-        final AppContext middleContext = new AppContext();
-        await middleContext.runInZone(() async {
-          final AppContext innerContext = new AppContext();
-          await innerContext.runInZone(() async {
-            expect(innerContext.getVariable(String), isNull);
-          });
+    group('run', () {
+      test('returns the value returned by body', () async {
+        expect(context.run<int>(body: () => 123), 123);
+        expect(context.run<String>(body: () => 'value'), 'value');
+        expect(await context.run<Future<int>>(body: () async => 456), 456);
+      });
+
+      test('passes name to child context', () {
+        context.run(name: 'child', body: () {
+          expect(context.name, 'child');
         });
       });
-    });
 
-    test('fireAndForgetNestedZones', () async {
-      final AppContext outerContext = new AppContext();
-      outerContext.runInZone(() async {
-        final AppContext middleContext = new AppContext();
-        middleContext.runInZone(() async {
-          final AppContext innerContext = new AppContext();
-          innerContext.runInZone(() async {
-            expect(innerContext.getVariable(String), isNull);
-          });
+      group('fallbacks', () {
+        bool called;
+
+        setUp(() {
+          called = false;
+        });
+
+        test('are applied after parent context is consulted', () {
+          final String value = context.run<String>(
+            body: () {
+              return context.run<String>(
+                body: () {
+                  called = true;
+                  return context[String];
+                },
+                fallbacks: <Type, Generator>{
+                  String: () => 'child',
+                },
+              );
+            },
+          );
+          expect(called, isTrue);
+          expect(value, 'child');
```

Comment:
Did you mean to put `called = true` into a parent generator? I can't tell which check shows that the parent was consulted.

Reply:
No, this just exists to make sure that `context.run` didn't skip the body entirely.  I originally had a "parent consulted" check, but that can't work, because if the parent is consulted, it means it had a generator, so the fallback would never be consulted...

Reply:
Ah, got it. I was confused by the test description. I thought the test somehow tracked when the fallbacks are consulted relative to the parent.